Sources/Logging/LoggerWithSource.swift

```diff
@@ -0,0 +1,273 @@
+    @inlinable
+    public func log(level: Logger.Level,
+                    _ message: @autoclosure () -> Logger.Message,
+                    metadata: @autoclosure () -> Logger.Metadata? = nil,
+                    file: String = #file, function: String = #function, line: UInt = #line) {
+        self.logger.log(level: level,
+                        message(),
+                        metadata: metadata(),
+                        source: self.source,
+                        file: file, function: function, line: line)
+    }
+
+    /// Add, change, or remove a logging metadata item.
+    ///
+    /// - note: Logging metadata behaves as a value that means a change to the logging metadata will only affect the
+    ///         very `Logger` it was changed on.
+    @inlinable
+    public subscript(metadataKey metadataKey: String) -> Logger.Metadata.Value? {
+        get {
+            return self.logger[metadataKey: metadataKey]
+        }
+        set {
+            self.logger[metadataKey: metadataKey] = newValue
+        }
+    }
+
+    /// Get or set the log level configured for this `Logger`.
+    ///
+    /// - note: `Logger`s treat `logLevel` as a value. This means that a change in `logLevel` will only affect this
+    ///         very `Logger`. It it acceptable for logging backends to have some form of global log level override
+    ///         that affects multiple or even all loggers. This means a change in `logLevel` to one `Logger` might in
+    ///         certain cases have no effect.
+    @inlinable
+    public var logLevel: Logger.Level {
+        get {
+            return self.logger.logLevel
+        }
+        set {
+            self.logger.logLevel = newValue
+        }
+    }
+}
+
+extension LoggerWithSource {
+    /// Log a message passing with the `Logger.Level.trace` log level.
+    ///
+    /// If `.trace` is at least as severe as the `Logger`'s `logLevel`, it will be logged,
+    /// otherwise nothing will happen.
+    ///
+    /// - parameters:
+    ///    - message: The message to be logged. `message` can be used with any string interpolation literal.
+    ///    - metadata: One-off metadata to attach to this log message
+    ///    - file: The file this log message originates from (there's usually no need to pass it explicitly as it
+    ///            defaults to `#file`).
+    ///    - function: The function this log message originates from (there's usually no need to pass it explicitly as
+    ///                it defaults to `#function`).
+    ///    - line: The line this log message originates from (there's usually no need to pass it explicitly as it
+    ///            defaults to `#line`).
+    @inlinable
+    public func trace(_ message: @autoclosure () -> Logger.Message,
+                      metadata: @autoclosure () -> Logger.Metadata? = nil,
+                      file: String = #file, function: String = #function, line: UInt = #line) {
+        self.logger.trace(message(),
```

Comment:
I'm not sure about this, but I think `LoggerWithSource.trace` should delegate to `LoggerWithSource.log`, and not to `Logger.trace`. Otherwise `LoggerWithSource` seems inconsistent with `Logger`, and the reason for `LoggerWithSource.log` is unclear. WDYT?

Reply:
@avolokhov From the generated code's perspective this is indistinguishable so we can change at any point in time.

The reason I did add `log` with `LoggerWithSource` is that some people may be using `log(level: .trace, ...)` instead of `.trace` and I want it to be compatible.